sql/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/streaming/TransformWithStateConnectSuite.scala

```diff
@@ -0,0 +1,489 @@
+
+      withTempPath { dir =>
+        val path = dir.getCanonicalPath
+        testData
+          .toDS()
+          .toDF("key", "value")
+          .repartition(3)
+          .write
+          .parquet(path)
+
+        val testSchema =
+          StructType(Array(StructField("key", StringType), StructField("value", StringType)))
+
+        val initDf = Seq(("init_1", "40.0", "a"), ("init_2", "100.0", "b"))
+          .toDS()
+          .groupByKey(x => x._3)
+          .mapValues(x => x)
+
+        val q = spark.readStream
+          .schema(testSchema)
+          .option("maxFilesPerTrigger", 1)
+          .parquet(path)
+          .as[(String, String)]
+          .groupByKey(x => x._1)
+          .transformWithState(
+            new TestInitialStatefulProcessor(),
+            TimeMode.None(),
+            OutputMode.Update(),
+            initialState = initDf)
+          .writeStream
+          .format("memory")
+          .queryName("my_sink")
+          .start()
+
+        try {
+          q.processAllAvailable()
+          eventually(timeout(30.seconds)) {
+            checkDatasetUnorderly(
+              spark.table("my_sink").toDF().as[(String, String)],
+              ("a", "2"),
+              ("a", "3"),
+              ("b", "2"))
+          }
+        } finally {
+          q.stop()
+          spark.sql("DROP TABLE IF EXISTS my_sink")
+        }
+      }
+    }
+  }
+
+  test("transformWithState - streaming with chaining of operators") {
+    withSQLConf(twsAdditionalSQLConf: _*) {
+      val session: SparkSession = spark
+      import session.implicits._
+
+      def timestamp(num: Int): Timestamp = {
+        new Timestamp(num * 1000)
+      }
+
+      val checkResultFunc: (Dataset[Row], Long) => Unit = { (batchDF, batchId) =>
+        val realDf = batchDF.collect().toSet
+        if (batchId == 0) {
+          assert(realDf.isEmpty, s"BatchId: $batchId, RealDF: $realDf")
+        } else if (batchId == 1) {
+          // eviction watermark = 15 - 5 = 10 (max event time from batch 0),
+          // late event watermark = 0 (eviction event time from batch 0)
+          val expectedDF = Seq(Row(timestamp(10), 1L)).toSet
+          assert(
+            realDf == expectedDF,
+            s"BatchId: $batchId, expectedDf: $expectedDF, RealDF: $realDf")
+        } else if (batchId == 2) {
+          // eviction watermark = 25 - 5 = 20, late event watermark = 10;
+          // row with watermark=5<10 is dropped so it does not show up in the results;
+          // row with eventTime<=20 are finalized and emitted
+          val expectedDF = Seq(Row(timestamp(11), 1L), Row(timestamp(15), 1L)).toSet
+          assert(
+            realDf == expectedDF,
+            s"BatchId: $batchId, expectedDf: $expectedDF, RealDF: $realDf")
+        }
+      }
+
+      withTempPath { dir =>
+        val path = dir.getCanonicalPath
+        prepareInputData(path + "/text-test3.csv", Seq("a", "b"), Seq(10, 15))
+        Thread.sleep(2000)
```

Comment:
Why not apply the technique on changing last updated timestamp here as well? This will save us 4 seconds in wall clock.